diff -Naur libmodplug-0.8.8.5.old/src/load_abc.cpp libmodplug-0.8.8.5/src/load_abc.cpp
--- libmodplug-0.8.8.5.old/src/load_abc.cpp	2014-03-09 23:53:33.000000000 +0000
+++ libmodplug-0.8.8.5/src/load_abc.cpp	2018-03-28 17:40:55.225959286 +0100
@@ -3844,8 +3844,10 @@
 	mm.sz = dwMemLength;
 	mm.pos = 0;
 #endif
+#ifndef __SWITCH__
 	while( avoid_reentry ) sleep(1);
 	avoid_reentry = 1;
+#endif
 	pat_resetsmp();
 	pat_init_patnames();
 	m_nDefaultTempo = 0;
diff -Naur libmodplug-0.8.8.5.old/src/load_mid.cpp libmodplug-0.8.8.5/src/load_mid.cpp
--- libmodplug-0.8.8.5.old/src/load_mid.cpp	2014-02-04 06:14:00.000000000 +0000
+++ libmodplug-0.8.8.5/src/load_mid.cpp	2018-03-28 17:40:21.750185049 +0100
@@ -1423,8 +1423,10 @@
 	BYTE midibyte[2];
 	long metalen, delta;
 	BYTE *p;
+#ifndef __SWITCH__
 	while( avoid_reentry ) sleep(1);
 	avoid_reentry = 1;
+#endif
 #ifdef NEWMIKMOD
 	h->mmf = mmfile;
 #else
diff -Naur libmodplug-0.8.8.5.old/src/load_pat.cpp libmodplug-0.8.8.5/src/load_pat.cpp
--- libmodplug-0.8.8.5.old/src/load_pat.cpp	2014-02-09 23:43:58.000000000 +0000
+++ libmodplug-0.8.8.5/src/load_pat.cpp	2018-03-28 17:38:22.492762495 +0100
@@ -1453,8 +1453,10 @@
 	mm.pos = 0;
 	mm.error = 0;
 #endif
+#ifndef __SWITCH__
 	while( avoid_reentry ) sleep(1);
 	avoid_reentry = 1;
+#endif
 	pat_read_patname(h, mmfile);
 	h->samples = pat_read_numsmp(mmfile);
 	if( strlen(h->patname) )
