--- mesa-20.1.0-rc3/src/mapi/glapi/gen/gl_XML.py.orig	2020-05-13 18:41:14.886004200 +0100
+++ mesa-20.1.0-rc3/src/mapi/glapi/gen/gl_XML.py	2025-07-04 16:25:01.736192731 +0100
@@ -716,7 +716,7 @@
 
         parameters = []
         return_type = "void"
-        for child in element.getchildren():
+        for child in list(element):
             if child.tag == "return":
                 return_type = child.get( "type", "void" )
             elif child.tag == "param":
@@ -746,7 +746,7 @@
                 if param.is_image():
                     self.images.append( param )
 
-        if element.getchildren():
+        if list(element):
             self.initialized = 1
             self.entry_point_parameters[name] = parameters
         else:
@@ -876,7 +876,7 @@
 
 
     def process_OpenGLAPI(self, file_name, element):
-        for child in element.getchildren():
+        for child in list(element):
             if child.tag == "category":
                 self.process_category( child )
             elif child.tag == "OpenGLAPI":
@@ -896,7 +896,7 @@
         [cat_type, key] = classify_category(cat_name, cat_number)
         self.categories[cat_type][key] = [cat_name, cat_number]
 
-        for child in cat.getchildren():
+        for child in list(cat):
             if child.tag == "function":
                 func_name = real_function_name( child )
 
