diff -Naur libarchive-3.3.2.orig/libarchive/archive_read_disk_posix.c libarchive-3.3.2/libarchive/archive_read_disk_posix.c
--- libarchive-3.3.2.orig/libarchive/archive_read_disk_posix.c	2017-02-25 18:37:08.000000000 +0100
+++ libarchive-3.3.2/libarchive/archive_read_disk_posix.c	2022-05-12 13:29:38.214610720 +0200
@@ -100,8 +100,10 @@
 #include "archive_read_disk_private.h"
 
 #ifndef HAVE_FCHDIR
+#ifndef __SWITCH__
 #error fchdir function required.
 #endif
+#endif
 #ifndef O_BINARY
 #define O_BINARY	0
 #endif
diff -Naur libarchive-3.3.2.orig/libarchive/archive_read_support_filter_program.c libarchive-3.3.2/libarchive/archive_read_support_filter_program.c
--- libarchive-3.3.2.orig/libarchive/archive_read_support_filter_program.c	2017-02-05 20:28:19.000000000 +0100
+++ libarchive-3.3.2/libarchive/archive_read_support_filter_program.c	2022-05-12 13:29:38.214610720 +0200
@@ -58,7 +58,13 @@
 #include "archive_read_private.h"
 #include "filter_fork.h"
 
-
+#ifdef __SWITCH__
+int
+__archive_read_program(struct archive_read_filter *self, const char *cmd)
+{
+	return (ARCHIVE_OK);
+}
+#else
 #if ARCHIVE_VERSION_NUMBER < 4000000
 /* Deprecated; remove in libarchive 4.0 */
 int
@@ -516,3 +522,4 @@
 
 	return (e);
 }
+#endif
diff -Naur libarchive-3.3.2.orig/libarchive/archive_write_set_format_iso9660.c libarchive-3.3.2/libarchive/archive_write_set_format_iso9660.c
--- libarchive-3.3.2.orig/libarchive/archive_write_set_format_iso9660.c	2017-02-25 18:37:08.000000000 +0100
+++ libarchive-3.3.2/libarchive/archive_write_set_format_iso9660.c	2022-05-12 13:29:38.214610720 +0200
@@ -2178,8 +2178,10 @@
 	strncpy(system_id, "Windows", size-1);
 	system_id[size-1] = '\0';
 #else
+#ifndef __SWITCH__
 #error no way to get the system identifier on your platform.
 #endif
+#endif
 }
 
 static void
@@ -2505,7 +2507,7 @@
 
 #if defined(HAVE__GET_TIMEZONE)
 	_get_timezone(&offset);
-#elif defined(__CYGWIN__) || defined(__MINGW32__) || defined(__BORLANDC__)
+#elif defined(__CYGWIN__) || defined(__MINGW32__) || defined(__BORLANDC__) || defined(__SWITCH__)
 	offset = _timezone;
 #else
 	offset = timezone;
diff -Naur libarchive-3.3.2.orig/libarchive_fe/passphrase.c libarchive-3.3.2/libarchive_fe/passphrase.c
--- libarchive-3.3.2.orig/libarchive_fe/passphrase.c	2016-12-03 07:52:27.000000000 +0100
+++ libarchive-3.3.2/libarchive_fe/passphrase.c	2022-05-12 13:29:27.558542060 +0200
@@ -129,9 +129,12 @@
 #endif
 #include <signal.h>
 #include <string.h>
+#ifndef __SWITCH__
 #include <termios.h>
+#endif
 #include <unistd.h>
 
+#ifndef __SWITCH__
 #ifndef _PATH_TTY
 #define _PATH_TTY "/dev/tty"
 #endif
@@ -303,6 +306,13 @@
 		errno = save_errno;
 	return(nr == -1 ? NULL : buf);
 }
+#else
+static char *
+readpassphrase(const char *prompt, char *buf, size_t bufsiz, int flags)
+{
+  return NULL;
+}
+#endif /* __SWITCH__ */
 #endif /* _WIN32 && !__CYGWIN__ */
 #endif /* HAVE_READPASSPHRASE */
 
