diff --git a/Makerules b/Makerules
index b96fc7138..91b853208 100644
--- a/Makerules
+++ b/Makerules
@@ -201,3 +201,18 @@ ifeq "$(OS)" "ios"
   LD = xcrun ld
   RANLIB = xcrun ranlib
 endif
+
+ifeq "$(OS)" "switch"
+  CC = $(TOOL_PREFIX)gcc
+  CXX = $(TOOL_PREFIX)g++
+  AR = $(TOOL_PREFIX)ar
+  LD = $(TOOL_PREFIX)ld
+  RANLIB = $(TOOL_PREFIX)ranlib
+  XCFLAGS = $(CPPFLAGS) '-Dtimegm(x)=-1'
+  USE_SYSTEM_FREETYPE = yes
+  SYS_FREETYPE_CFLAGS = $(shell freetype-config --cflags)
+  SYS_FREETYPE_LIBS = $(shell freetype-config --libs)
+  USE_SYSTEM_LIBJPEG = yes
+  USE_SYSTEM_ZLIB = yes
+endif
+
diff --git a/source/fitz/string.c b/source/fitz/string.c
index 2918c3f16..c8b709c2f 100644
--- a/source/fitz/string.c
+++ b/source/fitz/string.c
@@ -158,7 +158,11 @@ char *fz_realpath(const char *path, char buf[PATH_MAX])
 
 char *fz_realpath(const char *path, char buf[PATH_MAX])
 {
+#ifdef __SWITCH__
+	return NULL;
+#else
 	return realpath(path, buf);
+#endif
 }
 
 #endif
Submodule thirdparty/harfbuzz contains modified content
diff --git a/thirdparty/harfbuzz/src/hb.hh b/thirdparty/harfbuzz/src/hb.hh
index fcbd3305..ff36ef63 100644
--- a/thirdparty/harfbuzz/src/hb.hh
+++ b/thirdparty/harfbuzz/src/hb.hh
@@ -182,6 +182,7 @@
 #include <assert.h>
 #include <stdio.h>
 #include <stdarg.h>
+#include <climits>
 
 #if (defined(_MSC_VER) && _MSC_VER >= 1500) || defined(__MINGW32__)
 #ifdef __MINGW32_VERSION
Submodule thirdparty/lcms2 05db680de...29b0d7953 (commits not present)
Submodule thirdparty/libjpeg 9a532f2d5...cae42f6ea (commits not present)
