diff -urN a/audio/out/ao.c b/audio/out/ao.c
--- a/audio/out/ao.c	2025-12-21 20:08:20.000000000 +0100
+++ b/audio/out/ao.c	2025-12-22 15:59:01.661990655 +0100
@@ -54,6 +54,7 @@
 extern const struct ao_driver audio_out_pcm;
 extern const struct ao_driver audio_out_lavc;
 extern const struct ao_driver audio_out_sdl;
+extern const struct ao_driver audio_out_hos;
 extern const struct ao_driver audio_out_aaudio;
 
 static const struct ao_driver * const audio_out_drivers[] = {
@@ -104,6 +105,9 @@
 #if HAVE_SNDIO
     &audio_out_sndio,
 #endif
+#if HAVE_HOS_AUDIO
+    &audio_out_hos,
+#endif
     &audio_out_null,
 #if HAVE_COREAUDIO
     &audio_out_coreaudio_exclusive,
diff -urN a/audio/out/ao_hos.c b/audio/out/ao_hos.c
--- a/audio/out/ao_hos.c	1970-01-01 01:00:00.000000000 +0100
+++ b/audio/out/ao_hos.c	2025-12-22 15:00:08.143704970 +0100
@@ -0,0 +1,298 @@
+/*
+ * audio output driver for Horizon OS using audren
+ * Copyright (c) 2024 averne <averne381@gmail.com>
+ *
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <switch.h>
+
+#include "config.h"
+#include "common/common.h"
+#include "common/msg.h"
+#include "options/m_option.h"
+#include "audio/format.h"
+#include "ao.h"
+#include "internal.h"
+
+#define MAX_CHANS 6 // 5.1
+#define MAX_BUF 16
+#define MAX_SAMPLES 32768
+
+struct priv {
+    AudioDriver driver;
+    int num_buffers;
+    int num_samples;
+
+    void *pool;
+    AudioDriverWaveBuf *buffers;
+
+    int cur_buf_idx;
+    uint32_t cur_queued_samples, total_queued_samples;
+};
+
+static const AudioRendererConfig ar_config = {
+    .output_rate     = AudioRendererOutputRate_48kHz,
+    .num_voices      = MAX_CHANS,
+    .num_effects     = 0,
+    .num_sinks       = 1,
+    .num_mix_objs    = 1,
+    .num_mix_buffers = MAX_CHANS,
+};
+
+static const uint8_t sink_channel_ids[] = { 0, 1, 2, 3, 4, 5 };
+
+static const struct mp_chmap possible_channel_layouts[] = {
+    {0},
+    MP_CHMAP_INIT_MONO,                 // mono
+    MP_CHMAP_INIT_STEREO,               // stereo
+    MP_CHMAP3(FL, FR, LFE),             // 2.1
+    MP_CHMAP4(FL, FR, BL, BR),          // 4.0
+    MP_CHMAP5(FL, FR, FC, BL, BR),      // 5.0
+    MP_CHMAP6(FL, FR, FC, LFE, BL, BR), // 5.1
+};
+
+static int init(struct ao *ao) {
+    struct priv *priv = ao->priv;
+
+    Result rc;
+
+    MP_VERBOSE(ao, "Initializing hos audio\n");
+
+    ao->format   = AF_FORMAT_S16; // Only format supported by audrv with Adpcm which mpv can't output
+    ao->channels = possible_channel_layouts[MPMIN(ao->channels.num, MAX_CHANS)];
+
+    rc = audrenInitialize(&ar_config);
+    if (R_FAILED(rc))
+        return -rc;
+
+    rc = audrvCreate(&priv->driver, &ar_config, MAX_CHANS);
+    if (R_FAILED(rc))
+        return -rc;
+
+    size_t mempool_size = MP_ALIGN_UP(priv->num_samples * ao->channels.num *
+        priv->num_buffers * sizeof(int16_t), AUDREN_MEMPOOL_ALIGNMENT);
+
+    priv->pool = aligned_alloc(AUDREN_MEMPOOL_ALIGNMENT, mempool_size);
+    if (!priv->pool)
+        return -1;
+
+    priv->buffers = talloc_array(priv, AudioDriverWaveBuf, priv->num_buffers);
+    for (int i = 0; i < priv->num_buffers; ++i) {
+        priv->buffers[i] = (AudioDriverWaveBuf){
+            .data_raw            = priv->pool,
+            .size                = mempool_size,
+            .start_sample_offset = priv->num_samples * i,
+            .end_sample_offset   = priv->num_samples * (i + 1),
+        };
+    }
+
+    int mpid = audrvMemPoolAdd(&priv->driver, priv->pool, mempool_size);
+    audrvMemPoolAttach(&priv->driver, mpid);
+
+    ao->device_buffer = priv->num_buffers * priv->num_samples;
+
+    audrvDeviceSinkAdd(&priv->driver, AUDREN_DEFAULT_DEVICE_NAME, MAX_CHANS, sink_channel_ids);
+
+    rc = audrenStartAudioRenderer();
+    if (R_FAILED(rc))
+        return -rc;
+
+    audrvVoiceInit(&priv->driver, 0, ao->channels.num, PcmFormat_Int16, ao->samplerate);
+    audrvVoiceSetDestinationMix(&priv->driver, 0, AUDREN_FINAL_MIX_ID);
+
+    for (int i = 0; i < ao->channels.num; ++i)
+        audrvVoiceSetMixFactor(&priv->driver, 0, 1.0f, ao->channels.speaker[i], ao->channels.speaker[i]);
+
+    return 0;
+}
+
+static void uninit(struct ao *ao) {
+    struct priv *priv = ao->priv;
+
+    MP_VERBOSE(ao, "Deinitializing hos audio\n");
+
+    audrvVoiceStop(&priv->driver, 0);
+    audrvUpdate(&priv->driver);
+
+    audrvClose(&priv->driver);
+    audrenExit();
+
+    free(priv->pool);
+}
+
+static void reset(struct ao *ao) {
+    struct priv *priv = ao->priv;
+
+    priv->cur_buf_idx = -1;
+    priv->cur_queued_samples = priv->total_queued_samples = 0;
+    audrvVoiceStop(&priv->driver, 0);
+    audrvUpdate(&priv->driver);
+}
+
+static bool set_pause(struct ao *ao, bool paused) {
+    struct priv *priv = ao->priv;
+
+    audrvVoiceSetPaused(&priv->driver, 0, paused);
+    return R_SUCCEEDED(audrvUpdate(&priv->driver));
+}
+
+static void start(struct ao *ao) {
+    struct priv *priv = ao->priv;
+
+    audrvVoiceStart(&priv->driver, 0);
+    audrvUpdate(&priv->driver);
+}
+
+static int find_free_wavebuf(struct priv *priv) {
+    for (int i = 0; i < priv->num_buffers; ++i) {
+        AudioDriverWaveBuf *buf = &priv->buffers[i];
+        if (buf->state == AudioDriverWaveBufState_Done ||
+                buf->state == AudioDriverWaveBufState_Free)
+            return i;
+    }
+    return -1;
+}
+
+static bool audio_write(struct ao *ao, void **data, int samples) {
+    struct priv *priv = ao->priv;
+
+    // We requested a linear format so there is only one buffer
+    uint8_t *dat = data[0];
+
+    while (samples) {
+        int idx = (priv->cur_buf_idx != -1) ? priv->cur_buf_idx : find_free_wavebuf(priv);
+        if (idx == -1)
+            return false;
+        priv->cur_buf_idx = idx;
+
+        AudioDriverWaveBuf *buf = &priv->buffers[idx];
+        uint8_t *buf_offset = (uint8_t *)buf->data_raw + (idx * priv->num_samples * ao->sstride);
+
+        size_t num_samples = MPMIN(samples, priv->num_samples - priv->cur_queued_samples);
+
+        memcpy(buf_offset + priv->cur_queued_samples * ao->sstride, dat, num_samples * ao->sstride);
+        priv->cur_queued_samples   += num_samples;
+        priv->total_queued_samples += num_samples;
+
+        dat     += num_samples * ao->sstride;
+        samples -= num_samples;
+
+        // Append buffer once it's full
+        if (priv->cur_queued_samples >= priv->num_samples) {
+            armDCacheFlush(buf_offset, priv->num_samples * ao->sstride);
+            audrvVoiceAddWaveBuf(&priv->driver, 0, buf);
+            audrvUpdate(&priv->driver);
+
+            priv->cur_buf_idx = -1, priv->cur_queued_samples = 0;
+        }
+    }
+
+    return true;
+}
+
+static void get_state(struct ao *ao, struct mp_pcm_state *state) {
+    struct priv *priv = ao->priv;
+
+    Result rc = audrvUpdate(&priv->driver);
+    if (R_FAILED(rc))
+        return;
+
+    state->free_samples = state->queued_samples = 0;
+    for (int i = 0; i < priv->num_buffers; ++i) {
+        AudioDriverWaveBuf *buf = &priv->buffers[i];
+        if (buf->state == AudioDriverWaveBufState_Free
+                || buf->state == AudioDriverWaveBufState_Done)
+            state->free_samples += priv->num_samples;
+    }
+
+    if (priv->cur_buf_idx != -1)
+        state->free_samples -= priv->num_samples - priv->cur_queued_samples;
+
+    state->queued_samples = priv->total_queued_samples -
+        audrvVoiceGetPlayedSampleCount(&priv->driver, 0);
+
+    state->delay = (double)state->queued_samples / ao->samplerate;
+
+    state->playing = audrvVoiceIsPlaying(&priv->driver, 0);
+}
+
+static int control(struct ao *ao, enum aocontrol cmd, void *arg) {
+    struct priv *priv = ao->priv;
+
+    int rc;
+
+    switch (cmd) {
+        case AOCONTROL_SET_MUTE:
+        case AOCONTROL_SET_VOLUME: {
+                float vol;
+                if (cmd == AOCONTROL_SET_MUTE) {
+                    bool in = *(bool *)arg;
+                    vol = !in;
+                } else {
+                    float *in = arg;
+                    vol = *in / 100.0f;
+                }
+
+                audrvMixSetVolume(&priv->driver, 0, vol);
+                rc = audrvUpdate(&priv->driver);
+            }
+            break;
+        case AOCONTROL_GET_MUTE:
+        case AOCONTROL_GET_VOLUME: {
+                rc = audrvUpdate(&priv->driver);
+                float vol = priv->driver.in_mixes[0].volume;
+                if (cmd == AOCONTROL_GET_MUTE) {
+                    bool *out = (bool *)arg;
+                    *out = !vol;
+                } else {
+                    float *out = arg;
+                    *out = vol * 100.0f;
+                }
+            }
+            break;
+        default:
+            return CONTROL_UNKNOWN;
+    }
+
+    return R_SUCCEEDED(rc) ? CONTROL_OK : CONTROL_ERROR;
+}
+
+#define OPT_BASE_STRUCT struct priv
+
+const struct ao_driver audio_out_hos = {
+    .description     = "HOS Audio",
+    .name            = "hos",
+    .init            = init,
+    .uninit          = uninit,
+    .reset           = reset,
+    .control         = control,
+    .set_pause       = set_pause,
+    .start           = start,
+    .write           = audio_write,
+    .get_state       = get_state,
+    .priv_size       = sizeof(struct priv),
+    .priv_defaults   = &(const struct priv){
+        .num_buffers = 4,
+        .num_samples = 8192,
+    },
+    .options         = (const struct m_option[]){
+        {"num-buffers", OPT_INT(num_buffers), M_RANGE(2,   MAX_BUF)},
+        {"num-samples", OPT_INT(num_samples), M_RANGE(256, MAX_SAMPLES)},
+        {0}
+    },
+    .options_prefix   = "ao-hos",
+};
diff -urN a/common/common.h b/common/common.h
--- a/common/common.h	2025-12-21 20:08:20.000000000 +0100
+++ b/common/common.h	2025-12-22 15:01:57.423602458 +0100
@@ -25,7 +25,7 @@
 
 #include "config.h"
 
-#if HAVE_POSIX || defined(__MINGW32__)
+#if HAVE_POSIX || defined(__MINGW32__) || defined(__SWITCH__)
 #include <strings.h>
 #include <unistd.h>
 #endif
diff -urN a/meson.build b/meson.build
--- a/meson.build	2025-12-21 20:08:20.000000000 +0100
+++ b/meson.build	2025-12-22 15:06:34.848250797 +0100
@@ -314,7 +314,8 @@
 
 darwin = host_machine.system() == 'darwin'
 win32 = host_machine.system() == 'cygwin' or host_machine.system() == 'windows'
-posix = not win32
+hos = host_machine.system() == 'horizon'
+posix = not (win32 or hos)
 
 features += {'darwin': darwin}
 features += {'posix': posix}
@@ -335,7 +336,7 @@
     flags += [mswin_flags, '-mwin32']
 endif
 
-if posix
+if posix or hos
     flags += ['-D_GNU_SOURCE']
 endif
 
@@ -355,6 +356,8 @@
     dependencies += libdl
 endif
 
+libnx = cc.find_library('nx', required: false)
+
 # C11 atomics are mandatory but linking to the library is not always required.
 dependencies += cc.find_library('atomic', required: false)
 
@@ -438,7 +441,18 @@
     main_fn_source = files('osdep/main-fn-unix.c')
 endif
 
-if posix and not darwin
+if hos
+    path_source  = files('osdep/path-unix.c')
+    timer_source = files('osdep/timer-linux.c')
+    sources += files('osdep/subprocess-dummy.c',
+                     'osdep/terminal-dummy.c',
+                     'osdep/main-fn-unix.c',
+                     'osdep/language-posix.c',
+                     'input/ipc-dummy.c',
+                     'sub/filter_regex.c')
+endif
+
+if posix and not darwin and not hos
     path_source = files('osdep/path-unix.c')
     timer_source = files('osdep/timer-linux.c')
 endif
@@ -914,6 +928,12 @@
     dependencies += sdl2
 endif
 
+features += {'hos-audio': hos and libnx.found()}
+if features['hos-audio']
+    sources += files('audio/out/ao_hos.c')
+    dependencies += libnx
+endif
+
 sndio = dependency('sndio', required: get_option('sndio'))
 features += {'sndio': sndio.found()}
 features += {'sndio-1-9': sndio.version().version_compare('>= 1.9.0')}
diff -urN a/osdep/io.c b/osdep/io.c
--- a/osdep/io.c	2025-12-21 20:08:20.000000000 +0100
+++ b/osdep/io.c	2025-12-22 15:35:04.149344793 +0100
@@ -63,6 +63,13 @@
 }
 
 #ifndef _WIN32
+#ifdef __SWITCH__
+int mp_make_cloexec_pipe(int pipes[2])
+{
+    pipes[0] = pipes[1] = -1;
+    return -1;
+}
+#else
 int mp_make_cloexec_pipe(int pipes[2])
 {
     if (pipe(pipes) != 0) {
@@ -74,6 +81,7 @@
         mp_set_cloexec(pipes[i]);
     return 0;
 }
+#endif
 
 // create a pipe, and set it to non-blocking (and also set FD_CLOEXEC)
 int mp_make_wakeup_pipe(int pipes[2])
diff -urN a/osdep/io.h b/osdep/io.h
--- a/osdep/io.h	2025-12-21 20:08:20.000000000 +0100
+++ b/osdep/io.h	2025-12-22 15:18:20.446078080 +0100
@@ -248,7 +248,11 @@
 
 #else /* __MINGW32__ */
 
+#ifdef __SWITCH__
+#include "switch/sys/mman.h"
+#else
 #include <sys/mman.h>
+#endif
 
 extern char **environ;
 
diff -urN a/osdep/switch/sys/mman.h b/osdep/switch/sys/mman.h
--- a/osdep/switch/sys/mman.h	1970-01-01 01:00:00.000000000 +0100
+++ b/osdep/switch/sys/mman.h	2025-12-22 15:14:18.823779144 +0100
@@ -0,0 +1,15 @@
+#ifndef _MEMMAP_H_
+#define _MEMMAP_H_
+
+#include <stdio.h>
+#include <stdint.h>
+
+#define PROT_READ       0x1
+#define PROT_WRITE      0x2
+#define MAP_SHARED      0x01
+#define MAP_FAILED      ((void *) -1)
+
+#define mmap(a, b, c, d, e, f) malloc(b)
+#define munmap(a, b) free(a)
+
+#endif
diff -urN a/player/lua/stats.lua b/player/lua/stats.lua
--- a/player/lua/stats.lua	2025-12-21 20:08:20.000000000 +0100
+++ b/player/lua/stats.lua	2025-12-22 17:07:49.277266416 +0100
@@ -52,7 +52,7 @@
     plot_bg_border_width = 1.25,
 
     -- Text style
-    font = "",
+    font = "nintendo_udsg-r_std_003",
     font_mono = "monospace",   -- monospaced digits are sufficient
     font_size = 20,
     font_color = "",
diff -urN a/sub/filter_regex.c b/sub/filter_regex.c
--- a/sub/filter_regex.c	2025-12-21 20:08:20.000000000 +0100
+++ b/sub/filter_regex.c	2025-12-22 15:19:17.952460158 +0100
@@ -1,5 +1,5 @@
-#include <regex.h>
 #include <sys/types.h>
+#include <regex.h>
 
 #include "common/common.h"
 #include "common/msg.h"
diff -urN a/sub/osd.c b/sub/osd.c
--- a/sub/osd.c	2025-12-21 20:08:20.000000000 +0100
+++ b/sub/osd.c	2025-12-22 15:21:04.487040006 +0100
@@ -80,7 +80,7 @@
     .opts = style_opts,
     .size = sizeof(struct osd_style_opts),
     .defaults = &(const struct osd_style_opts){
-        .font = "sans-serif",
+        .font = "nintendo_udsg-r_std_003",
         .font_size = 30,
         .color = {255, 255, 255, 255},
         .outline_color = {0, 0, 0, 255},
@@ -100,7 +100,7 @@
     .opts = style_opts,
     .size = sizeof(struct osd_style_opts),
     .defaults = &(const struct osd_style_opts){
-        .font = "sans-serif",
+        .font = "nintendo_udsg-r_std_003",
         .font_size = 38,
         .color = {255, 255, 255, 255},
         .outline_color = {0, 0, 0, 255},
diff -urN a/sub/sd_ass.c b/sub/sd_ass.c
--- a/sub/sd_ass.c	2025-12-21 20:08:20.000000000 +0100
+++ b/sub/sd_ass.c	2025-12-22 15:22:18.877360486 +0100
@@ -21,6 +21,10 @@
 #include <math.h>
 #include <limits.h>
 
+#ifdef __SWITCH__
+#include <switch.h>
+#endif
+
 #include <libavutil/common.h>
 #include <ass/ass.h>
 
@@ -240,6 +244,18 @@
     } else {
         ctx->ass_renderer = ass_renderer_init(ctx->ass_library);
 
+#ifdef __SWITCH__
+        if (R_SUCCEEDED(plInitialize(PlServiceType_User))) {
+            PlFontData font;
+            Result rc = plGetSharedFontByType(&font, PlSharedFontType_Standard);
+            if (R_SUCCEEDED(rc))
+                ass_add_font(ctx->ass_library, "nintendo_udsg-r_std_003", font.address, font.size);
+            else
+                MP_ERR(sd, "Failed to add font from pl: %#x\n", rc);
+            plExit();
+        }
+#endif
+
         mp_ass_configure_fonts(ctx->ass_renderer, sd->opts->sub_style,
                                sd->global, sd->log);
     }
diff -urN a/video/mp_image.c b/video/mp_image.c
--- a/video/mp_image.c	2025-12-21 20:08:20.000000000 +0100
+++ b/video/mp_image.c	2025-12-22 15:22:51.902483440 +0100
@@ -175,7 +175,7 @@
         return false;
 
     // Note: mp_image_pool assumes this creates only 1 AVBufferRef.
-    mpi->bufs[0] = av_buffer_alloc(size + align);
+    mpi->bufs[0] = av_buffer_aligned_alloc(size + align, 0x100); // Alignment required for plane offsets
     if (!mpi->bufs[0])
         return false;
 
diff -urN a/video/mp_image.h b/video/mp_image.h
--- a/video/mp_image.h	2025-12-21 20:08:20.000000000 +0100
+++ b/video/mp_image.h	2025-12-22 15:23:25.994599008 +0100
@@ -32,7 +32,7 @@
 // libraries except libavcodec don't really know what alignment they want.
 // Things will randomly crash or get slower if the alignment is not satisfied.
 // Whatever. This value should be pretty safe with current CPU architectures.
-#define MP_IMAGE_BYTE_ALIGN 64
+#define MP_IMAGE_BYTE_ALIGN 256 // VIC alignement
 
 #define MP_IMGFIELD_TOP_FIRST 0x02
 #define MP_IMGFIELD_REPEAT_FIRST 0x04
