diff -NBaur mbedtls-2.14.1/include/mbedtls/config.h mbedtls-2.14.1.switch/include/mbedtls/config.h
--- mbedtls-2.14.1/include/mbedtls/config.h	2018-12-02 13:20:52.000000000 +0000
+++ mbedtls-2.14.1.switch/include/mbedtls/config.h	2018-12-23 18:49:07.178930622 +0000
@@ -476,7 +476,7 @@
  *
  * Uncomment to use your own hardware entropy collector.
  */
-//#define MBEDTLS_ENTROPY_HARDWARE_ALT
+#define MBEDTLS_ENTROPY_HARDWARE_ALT
 
 /**
  * \def MBEDTLS_AES_ROM_TABLES
@@ -1057,7 +1057,7 @@
  *
  * Uncomment this macro to disable the built-in platform entropy functions.
  */
-//#define MBEDTLS_NO_PLATFORM_ENTROPY
+#define MBEDTLS_NO_PLATFORM_ENTROPY
 
 /**
  * \def MBEDTLS_ENTROPY_FORCE_SHA256
@@ -1175,7 +1175,7 @@
  *
  * Enable the checkup functions (*_self_test).
  */
-#define MBEDTLS_SELF_TEST
+//#define MBEDTLS_SELF_TEST
 
 /**
  * \def MBEDTLS_SHA256_SMALLER
@@ -2795,7 +2795,7 @@
  *
  * This module is used by the HAVEGE random number generator.
  */
-#define MBEDTLS_TIMING_C
+//#define MBEDTLS_TIMING_C
 
 /**
  * \def MBEDTLS_VERSION_C
diff -NBaur mbedtls-2.14.1/library/entropy.c mbedtls-2.14.1.switch/library/entropy.c
--- mbedtls-2.14.1/library/entropy.c	2018-12-02 13:20:52.000000000 +0000
+++ mbedtls-2.14.1.switch/library/entropy.c	2018-12-23 18:49:07.178930622 +0000
@@ -119,6 +119,19 @@
 #endif /* MBEDTLS_NO_DEFAULT_ENTROPY_SOURCES */
 }
 
+#ifdef __SWITCH__
+    #include <switch.h>
+    int mbedtls_hardware_poll( void *data,
+                               unsigned char *output, size_t len, size_t *olen )
+    {
+        (void)(data);
+        csrngGetRandomBytes(output, len);
+        if(olen)
+            *olen = len;
+        return 0;
+    }
+#endif
+
 void mbedtls_entropy_free( mbedtls_entropy_context *ctx )
 {
 #if defined(MBEDTLS_HAVEGE_C)
diff -NBaur mbedtls-2.14.1/library/net_sockets.c mbedtls-2.14.1.switch/library/net_sockets.c
--- mbedtls-2.14.1/library/net_sockets.c	2018-12-02 13:20:52.000000000 +0000
+++ mbedtls-2.14.1.switch/library/net_sockets.c	2018-12-23 18:54:12.216006757 +0000
@@ -34,7 +34,7 @@
 
 #if !defined(unix) && !defined(__unix__) && !defined(__unix) && \
     !defined(__APPLE__) && !defined(_WIN32) && !defined(__QNXNTO__) && \
-    !defined(__HAIKU__)
+    !defined(__HAIKU__) && !defined(__SWITCH__)
 #error "This module only works on Unix and Windows, see MBEDTLS_NET_C in config.h"
 #endif
 
@@ -409,7 +409,7 @@
 
             memcpy( client_ip, &addr4->sin_addr.s_addr, *ip_len );
         }
-        else
+/*        else
         {
             struct sockaddr_in6 *addr6 = (struct sockaddr_in6 *) &client_addr;
             *ip_len = sizeof( addr6->sin6_addr.s6_addr );
@@ -419,7 +419,7 @@
 
             memcpy( client_ip, &addr6->sin6_addr.s6_addr, *ip_len);
         }
-    }
+*/    }
 
     return( 0 );
 }
