--- bullet3-2.87/CMakeLists.txt	2017-09-30 00:20:39.000000000 +0100
+++ bullet3-2.87-3ds/CMakeLists.txt	2018-11-20 01:14:42.028362942 +0000
@@ -282,15 +282,17 @@
 # This is the shortcut to finding GLU, GLUT and OpenGL if they are properly installed on your system
 # This should be the case.
 
-FIND_PACKAGE(OpenGL)
-IF (OPENGL_FOUND)
+IF(NOT SWITCH_LIBNX)
+  FIND_PACKAGE(OpenGL)
+  IF (OPENGL_FOUND)
 	MESSAGE("OPENGL FOUND")
 	MESSAGE(${OPENGL_LIBRARIES})
-ELSE (OPENGL_FOUND)
+  ELSE (OPENGL_FOUND)
 	MESSAGE("OPENGL NOT FOUND")
 	SET(OPENGL_gl_LIBRARY opengl32)
 	SET(OPENGL_glu_LIBRARY glu32)
-ENDIF (OPENGL_FOUND)
+  ENDIF (OPENGL_FOUND)
+ENDIF(NOT SWITCH_LIBNX)
 
 
 #FIND_PACKAGE(GLU)
@@ -303,6 +305,7 @@
 OPTION(BUILD_BULLET3 "Set when you want to build Bullet 3" ON)
 # Optional Python configuration
 # builds pybullet automatically if all the requirements are met
+IF(NOT SWITCH_LIBNX)
 SET(PYTHON_VERSION_PYBULLET "" CACHE STRING "Python version pybullet will use.")
 SET(Python_ADDITIONAL_VERSIONS 3 3.6 3.5 3.4 3.3 3.2 3.1 3.0 2.7 2.7.12 2.7.10 2.7.3 )
 SET_PROPERTY(CACHE PYTHON_VERSION_PYBULLET PROPERTY STRINGS ${Python_ADDITIONAL_VERSIONS})
@@ -320,6 +323,7 @@
 	SET(DEFAULT_BUILD_PYBULLET ON)
 ENDIF(PYTHONLIBS_FOUND)
 OPTION(BUILD_PYBULLET "Set when you want to build pybullet (Python bindings for Bullet)" ${DEFAULT_BUILD_PYBULLET})
+ENDIF(NOT SWITCH_LIBNX)
 
 OPTION(BUILD_ENET "Set when you want to build apps with enet UDP networking support" ON)
 OPTION(BUILD_CLSOCKET "Set when you want to build apps with enet TCP networking support" ON)
