diff --git a/Makefile.in b/Makefile.in
index 3d10565..833ef35 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -47,7 +47,7 @@ LT_RELEASE  = @LT_RELEASE@
 LT_REVISION = @LT_REVISION@
 LT_LDFLAGS  = -no-undefined -rpath $(libdir) -release $(LT_RELEASE) -version-info $(LT_CURRENT):$(LT_REVISION):$(LT_AGE)
 
-all: $(srcdir)/configure Makefile $(objects) $(objects)/$(TARGET) $(objects)/playwave$(EXE) $(objects)/playmus$(EXE)
+all: $(srcdir)/configure Makefile $(objects) $(objects)/$(TARGET)
 
 $(srcdir)/configure: $(srcdir)/configure.in
 	@echo "Warning, configure.in is out of date"
@@ -60,18 +60,12 @@ Makefile: $(srcdir)/Makefile.in
 $(objects):
 	$(SHELL) $(auxdir)/mkinstalldirs $@
 
-.PHONY: all install install-hdrs install-lib install-bin uninstall uninstall-hdrs uninstall-lib uninstall-bin clean distclean dist
+.PHONY: all install install-hdrs install-lib uninstall uninstall-hdrs uninstall-lib clean distclean dist
 
 $(objects)/$(TARGET): $(OBJECTS) $(VERSION_OBJECTS)
 	$(LIBTOOL) --mode=link $(CC) -o $@ $(OBJECTS) $(VERSION_OBJECTS) $(LDFLAGS) $(EXTRA_LDFLAGS) $(LT_LDFLAGS)
 
-$(objects)/playwave$(EXE): $(objects)/playwave.lo $(objects)/$(TARGET)
-	$(LIBTOOL) --mode=link $(CC) -o $@ $(objects)/playwave.lo $(SDL_CFLAGS) $(SDL_LIBS) $(objects)/$(TARGET)
-
-$(objects)/playmus$(EXE): $(objects)/playmus.lo $(objects)/$(TARGET)
-	$(LIBTOOL) --mode=link $(CC) -o $@ $(objects)/playmus.lo $(SDL_CFLAGS) $(SDL_LIBS) $(objects)/$(TARGET)
-
-install: all install-hdrs install-lib #install-bin
+install: all install-hdrs install-lib
 install-hdrs:
 	$(SHELL) $(auxdir)/mkinstalldirs $(includedir)/SDL
 	for src in $(srcdir)/SDL_mixer.h; do \
@@ -83,12 +77,8 @@ install-hdrs:
 install-lib: $(objects) $(objects)/$(TARGET)
 	$(SHELL) $(auxdir)/mkinstalldirs $(libdir)
 	$(LIBTOOL) --mode=install $(INSTALL) $(objects)/$(TARGET) $(libdir)/$(TARGET)
-install-bin:
-	$(SHELL) $(auxdir)/mkinstalldirs $(bindir)
-	$(LIBTOOL) --mode=install $(INSTALL) -m 755 $(objects)/playwave$(EXE) $(bindir)/playwave$(EXE)
-	$(LIBTOOL) --mode=install $(INSTALL) -m 755 $(objects)/playmus$(EXE) $(bindir)/playmus$(EXE)
 
-uninstall: uninstall-hdrs uninstall-lib uninstall-bin
+uninstall: uninstall-hdrs uninstall-lib
 uninstall-hdrs:
 	for src in $(srcdir)/SDL_mixer.h; do \
 	    file=`echo $$src | sed -e 's|^.*/||'`; \
@@ -99,9 +89,6 @@ uninstall-hdrs:
 	-rmdir $(libdir)/pkgconfig
 uninstall-lib:
 	$(LIBTOOL) --mode=uninstall rm -f $(libdir)/$(TARGET)
-uninstall-bin:
-	rm -f $(bindir)/playwave$(EXE)
-	rm -f $(bindir)/playmus$(EXE)
 
 clean:
 	rm -rf $(objects)
diff --git a/load_ogg.c b/load_ogg.c
index 2d0ee65..1e94fa6 100644
--- a/load_ogg.c
+++ b/load_ogg.c
@@ -105,7 +105,7 @@ SDL_AudioSpec *Mix_LoadOGG_RW (SDL_RWops *src, int freesrc,
     *audio_len = 0;
     memset(spec, '\0', sizeof (SDL_AudioSpec));
 
-    spec->format = AUDIO_S16;
+    spec->format = AUDIO_S16SYS;
     spec->channels = info->channels;
     spec->freq = info->rate;
     spec->samples = 4096; /* buffer size */
@@ -124,9 +124,9 @@ SDL_AudioSpec *Mix_LoadOGG_RW (SDL_RWops *src, int freesrc,
 	 read > 0;
 	 read = vorbis.ov_read(&vf, (char *)buf, to_read, &bitstream))
 #else
-    for (read = vorbis.ov_read(&vf, (char *)buf, to_read, 0/*LE*/, 2/*16bit*/, 1/*signed*/, &bitstream);
+    for (read = vorbis.ov_read(&vf, (char *)buf, to_read, SDL_BYTEORDER == SDL_BIG_ENDIAN/*BE?*/, 2/*16bit*/, 1/*signed*/, &bitstream);
          read > 0;
-         read = vorbis.ov_read(&vf, (char *)buf, to_read, 0, 2, 1, &bitstream))
+         read = vorbis.ov_read(&vf, (char *)buf, to_read, SDL_BYTEORDER == SDL_BIG_ENDIAN, 2, 1, &bitstream))
 #endif	 
     {
         if (read == OV_HOLE || read == OV_EBADLINK)
diff --git a/music_modplug.h b/music_modplug.h
index 92cbafd..555227b 100644
--- a/music_modplug.h
+++ b/music_modplug.h
@@ -1,6 +1,6 @@
 #ifdef MODPLUG_MUSIC
 
-#include "modplug.h"
+#include <libmodplug/modplug.h>
 #include "SDL_rwops.h"
 #include "SDL_audio.h"
 #include "SDL_mixer.h"
diff --git a/music_ogg.c b/music_ogg.c
index 18a094b..269a197 100644
--- a/music_ogg.c
+++ b/music_ogg.c
@@ -135,7 +135,7 @@ static void OGG_getsome(OGG_music *music)
 #ifdef OGG_USE_TREMOR
 	len = vorbis.ov_read(&music->vf, data, sizeof(data), &section);
 #else
-	len = vorbis.ov_read(&music->vf, data, sizeof(data), 0, 2, 1, &section);
+	len = vorbis.ov_read(&music->vf, data, sizeof(data), SDL_BYTEORDER == SDL_BIG_ENDIAN, 2, 1, &section);
 #endif
 	if ( len <= 0 ) {
 		if ( len == 0 ) {
@@ -148,7 +148,7 @@ static void OGG_getsome(OGG_music *music)
 		vorbis_info *vi;
 
 		vi = vorbis.ov_info(&music->vf, -1);
-		SDL_BuildAudioCVT(cvt, AUDIO_S16, vi->channels, vi->rate,
+		SDL_BuildAudioCVT(cvt, AUDIO_S16SYS, vi->channels, vi->rate,
 		                       mixer.format,mixer.channels,mixer.freq);
 		if ( cvt->buf ) {
 			SDL_free(cvt->buf);
