diff --git a/configure.ac b/configure.ac
index 6d9b04b..e35d8bd 100644
--- a/configure.ac
+++ b/configure.ac
@@ -61,6 +61,12 @@ AC_ARG_ENABLE(osmesa,
               [OSMESA="$enableval"],
               [OSMESA=no])
 
+AC_ARG_ENABLE(egl,
+              AS_HELP_STRING([--enable-egl],
+                [Enable use of libEGL instead of libGL]),
+              [EGL="$enableval"],
+              [EGL=no])
+
 dnl Get the pkg-config definitions for libGL/OSMesa.  We include a fallback
 dnl path for implementations that don't provide a .pc file
 if test "x$OSMESA" = "xyes"; then
@@ -71,17 +77,30 @@ if test "x$OSMESA" = "xyes"; then
                    [OSMESA_LIBS=-lOSMesa],
                    AC_MSG_ERROR([OSMesa required]))
           ])
-else
-    GLU_REQUIRES="gl"
-    PKG_CHECK_MODULES(GL, [gl], [], [
-          AC_CHECK_HEADER([GL/gl.h],
-                  [],
-                  AC_MSG_ERROR([GL not found]))
-          AC_CHECK_LIB([GL],
-                   [glBegin],
-                   [GL_LIBS=-lGL],
-                   AC_MSG_ERROR([GL required]))
-          ])
+else 
+    if test "x$EGL" = "xyes"; then
+        GLU_REQUIRES="egl"
+        PKG_CHECK_MODULES(EGL, [egl], [], [
+            AC_CHECK_HEADER([GL/gl.h],
+                    [],
+                    AC_MSG_ERROR([EGL not found]))
+            AC_CHECK_LIB([EGL],
+                    [glBegin],
+                    [GL_LIBS=-lEGL],
+                    AC_MSG_ERROR([EGL required]))
+            ])
+    else 
+        GLU_REQUIRES="gl"
+        PKG_CHECK_MODULES(GL, [gl], [], [
+            AC_CHECK_HEADER([GL/gl.h],
+                    [],
+                    AC_MSG_ERROR([GL not found]))
+            AC_CHECK_LIB([GL],
+                    [glBegin],
+                    [GL_LIBS=-lGL],
+                    AC_MSG_ERROR([GL required]))
+            ])
+    fi
 fi
 AC_SUBST([GLU_REQUIRES])
 
