diff -NBaur libgd-2.3.3.orig/src/gd_errors.h libgd-2.3.3/src/gd_errors.h
--- libgd-2.3.3.orig/src/gd_errors.h	2021-09-11 05:41:44.000000000 +0100
+++ libgd-2.3.3/src/gd_errors.h	2023-05-01 10:58:32.797327239 +0100
@@ -1,9 +1,6 @@
 #ifndef GD_ERRORS_H
 #define GD_ERRORS_H
 
-#ifndef _WIN32
-# include <syslog.h>
-#else
 /*
  * priorities/facilities are encoded into a single 32-bit quantity, where the
  * bottom 3 bits are the priority (0-7) and the top 28 bits are the facility
@@ -21,7 +18,6 @@
 # define LOG_NOTICE      5       /* normal but significant condition */
 # define LOG_INFO        6       /* informational */
 # define LOG_DEBUG       7       /* debug-level messages */
-#endif
 
 /*
 LOG_EMERG      system is unusable
diff -NBaur libgd-2.3.3.orig/src/Makefile.am libgd-2.3.3/src/Makefile.am
--- libgd-2.3.3.orig/src/Makefile.am	2021-09-11 05:43:11.000000000 +0100
+++ libgd-2.3.3/src/Makefile.am	2023-05-01 10:59:31.674301747 +0100
@@ -1,44 +1,7 @@
 ## Process this file with automake to produce Makefile.in -*-Makefile-*-
 
-bin_PROGRAMS = gdcmpgif
-check_PROGRAMS = gifanimtest
-
-if HAVE_LIBPNG
-bin_PROGRAMS += gdtopng pngtogd webpng
-check_PROGRAMS += circletexttest fontsizetest fontwheeltest gddemo testac
-if HAVE_LIBZ
-bin_PROGRAMS += gdparttopng gd2topng pngtogd2
-check_PROGRAMS += gdtest
-endif
-endif
-
-if HAVE_LIBJPEG
-if !HAVE_LIBPNG
-check_PROGRAMS += fontsizetest fontwheeltest
-endif
-endif
-
-if HAVE_LIBFONTCONFIG
-check_PROGRAMS += fontconfigtest
-endif
-
-if HAVE_LIBFREETYPE
-if HAVE_LIBJPEG
-bin_PROGRAMS += annotate
-endif
-check_PROGRAMS += gdtestft testtr
-endif
-
-if HAVE_LIBZ
-bin_PROGRAMS += gd2copypal gd2togif giftogd2
-endif
-
 bin_SCRIPTS = bdftogd
 
-if HAVE_LIBZ
-check_PROGRAMS += gd2time
-endif
-
 EXTRA_DIST = \
 	bdftogd \
 	demoin.png \
