--- Makefile.in.orig	2023-02-07 00:43:37.000000000 +0000
+++ Makefile.in	2023-09-11 23:37:03.634801000 +0000
@@ -93,7 +93,6 @@
 host_triplet = @host@
 @USE_VERSION_RC_FALSE@libSDL2_image_la_DEPENDENCIES =  \
 @USE_VERSION_RC_FALSE@	$(am__DEPENDENCIES_1)
-noinst_PROGRAMS = showimage$(EXEEXT) showanim$(EXEEXT)
 @BUILD_TESTS_TRUE@am__append_1 = test
 subdir = .
 ACLOCAL_M4 = $(top_srcdir)/aclocal.m4
@@ -171,12 +170,6 @@
 libSDL2_image_la_LINK = $(LIBTOOL) $(AM_V_lt) $(AM_LIBTOOLFLAGS) \
 	$(LIBTOOLFLAGS) --mode=link $(OBJCLD) $(AM_OBJCFLAGS) \
 	$(OBJCFLAGS) $(libSDL2_image_la_LDFLAGS) $(LDFLAGS) -o $@
-showanim_SOURCES = showanim.c
-showanim_OBJECTS = showanim.$(OBJEXT)
-showanim_DEPENDENCIES = libSDL2_image.la
-showimage_SOURCES = showimage.c
-showimage_OBJECTS = showimage.$(OBJEXT)
-showimage_DEPENDENCIES = libSDL2_image.la
 AM_V_P = $(am__v_P_@AM_V@)
 am__v_P_ = $(am__v_P_@AM_DEFAULT_V@)
 am__v_P_0 = false
@@ -578,8 +571,6 @@
 @USE_VERSION_RC_TRUE@libSDL2_image_la_DEPENDENCIES = version.o
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = SDL2_image.pc
-showimage_LDADD = libSDL2_image.la
-showanim_LDADD = libSDL2_image.la
 SUBDIRS = . $(am__append_1)
 all: all-recursive
 
@@ -674,14 +665,6 @@
 libSDL2_image.la: $(libSDL2_image_la_OBJECTS) $(libSDL2_image_la_DEPENDENCIES) $(EXTRA_libSDL2_image_la_DEPENDENCIES) 
 	$(AM_V_OBJCLD)$(libSDL2_image_la_LINK) -rpath $(libdir) $(libSDL2_image_la_OBJECTS) $(libSDL2_image_la_LIBADD) $(LIBS)
 
-showanim$(EXEEXT): $(showanim_OBJECTS) $(showanim_DEPENDENCIES) $(EXTRA_showanim_DEPENDENCIES) 
-	@rm -f showanim$(EXEEXT)
-	$(AM_V_CCLD)$(LINK) $(showanim_OBJECTS) $(showanim_LDADD) $(LIBS)
-
-showimage$(EXEEXT): $(showimage_OBJECTS) $(showimage_DEPENDENCIES) $(EXTRA_showimage_DEPENDENCIES) 
-	@rm -f showimage$(EXEEXT)
-	$(AM_V_CCLD)$(LINK) $(showimage_OBJECTS) $(showimage_LDADD) $(LIBS)
-
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
 
