diff -NBaur ntfs-3g-2022.10.3.old/include/ntfs-3g/ntfstime.h ntfs-3g-2022.10.3.new/include/ntfs-3g/ntfstime.h
--- ntfs-3g-2022.10.3.old/include/ntfs-3g/ntfstime.h	2022-10-03 10:10:36.000000000 +0100
+++ ntfs-3g-2022.10.3.new/include/ntfs-3g/ntfstime.h	2024-02-23 17:32:46.837523414 +0000
@@ -36,6 +36,10 @@
 
 #include "types.h"
 
+#ifdef _SYS__TIMESPEC_H_
+#define __timespec_defined
+#endif
+
 /*
  * assume "struct timespec" is not defined if st_mtime is not defined
  */
diff -NBaur ntfs-3g-2022.10.3.old/libntfs-3g/ioctl.c ntfs-3g-2022.10.3.new/libntfs-3g/ioctl.c
--- ntfs-3g-2022.10.3.old/libntfs-3g/ioctl.c	2022-10-03 10:10:36.000000000 +0100
+++ ntfs-3g-2022.10.3.new/libntfs-3g/ioctl.c	2024-02-23 17:32:46.837523414 +0000
@@ -48,7 +48,9 @@
 #ifdef HAVE_LIMITS_H
 #include <limits.h>
 #endif
+#ifdef HAVE_SYSLOG_H
 #include <syslog.h>
+#endif
 #ifdef HAVE_SYS_TYPES_H
 #include <sys/types.h>
 #endif
diff -NBaur ntfs-3g-2022.10.3.old/libntfs-3g/Makefile.am ntfs-3g-2022.10.3.new/libntfs-3g/Makefile.am
--- ntfs-3g-2022.10.3.old/libntfs-3g/Makefile.am	2022-10-03 10:10:36.000000000 +0100
+++ ntfs-3g-2022.10.3.new/libntfs-3g/Makefile.am	2024-02-23 17:36:19.525460770 +0000
@@ -57,26 +57,26 @@
 
 # We may need to move .so files to root
 # And create ldscript or symbolic link from /usr
-install-exec-hook: install-rootlibLTLIBRARIES
-if INSTALL_LIBRARY
-	if [ ! "$(rootlibdir)" -ef "$(libdir)" ]; then \
-		$(MV) -f "$(DESTDIR)/$(libdir)"/libntfs-3g.so* "$(DESTDIR)/$(rootlibdir)";  \
-	fi
-if GENERATE_LDSCRIPT
-	if [ ! "$(rootlibdir)" -ef "$(libdir)" ]; then \
-		$(install_sh_PROGRAM) "libntfs-3g.script.so" "$(DESTDIR)/$(libdir)/libntfs-3g.so"; \
-	fi
-else
-	if [ ! "$(rootlibdir)" -ef "$(libdir)" ]; then \
-		$(LN_S) "$(rootlibdir)/libntfs-3g.so" "$(DESTDIR)/$(libdir)/libntfs-3g.so"; \
-	fi
-endif
-endif
+#install-exec-hook: install-rootlibLTLIBRARIES
+#if INSTALL_LIBRARY
+#	if [ ! "$(rootlibdir)" -ef "$(libdir)" ]; then \
+#		$(MV) -f "$(DESTDIR)/$(libdir)"/libntfs-3g.so* "$(DESTDIR)/$(rootlibdir)";  \
+#	fi
+#if GENERATE_LDSCRIPT
+#	if [ ! "$(rootlibdir)" -ef "$(libdir)" ]; then \
+#		$(install_sh_PROGRAM) "libntfs-3g.script.so" "$(DESTDIR)/$(libdir)/libntfs-3g.so"; \
+#	fi
+#else
+#	if [ ! "$(rootlibdir)" -ef "$(libdir)" ]; then \
+#		$(LN_S) "$(rootlibdir)/libntfs-3g.so" "$(DESTDIR)/$(libdir)/libntfs-3g.so"; \
+#	fi
+#endif
+#endif
 
-uninstall-local:
-if INSTALL_LIBRARY
-	$(RM) -f "$(DESTDIR)/$(rootlibdir)"/libntfs-3g.so*
-endif
+#uninstall-local:
+#if INSTALL_LIBRARY
+#	$(RM) -f "$(DESTDIR)/$(rootlibdir)"/libntfs-3g.so*
+#endif
 
 if ENABLE_NTFSPROGS
 libs:	$(lib_LTLIBRARIES)
