diff -Nbaur mbedtls-2.16.6/library/entropy.c mbedtls-2.16.6-new/library/entropy.c
--- mbedtls-2.16.6/library/entropy.c	2020-04-09 22:12:23.000000000 +0900
+++ mbedtls-2.16.6-new/library/entropy.c	2020-04-16 00:21:50.625234859 +0900
@@ -119,6 +119,20 @@
 #endif /* MBEDTLS_NO_DEFAULT_ENTROPY_SOURCES */
 }
 
+#define LIBCTRU
+#ifdef LIBCTRU
+    #include <3ds.h>
+    int mbedtls_hardware_poll( void *data,
+                               unsigned char *output, size_t len, size_t *olen )
+    {
+        (void)(data);
+        sslcGenerateRandomData(output, len);
+        if(olen)
+            *olen = len;
+        return 0;
+    }
+#endif
+
 void mbedtls_entropy_free( mbedtls_entropy_context *ctx )
 {
 #if defined(MBEDTLS_HAVEGE_C)
diff -Nbaur mbedtls-2.16.6/library/net_sockets.c mbedtls-2.16.6-new/library/net_sockets.c
--- mbedtls-2.16.6/library/net_sockets.c	2020-04-09 22:12:23.000000000 +0900
+++ mbedtls-2.16.6-new/library/net_sockets.c	2020-04-16 00:21:50.628568192 +0900
@@ -32,12 +32,6 @@
 
 #if defined(MBEDTLS_NET_C)
 
-#if !defined(unix) && !defined(__unix__) && !defined(__unix) && \
-    !defined(__APPLE__) && !defined(_WIN32) && !defined(__QNXNTO__) && \
-    !defined(__HAIKU__)
-#error "This module only works on Unix and Windows, see MBEDTLS_NET_C in config.h"
-#endif
-
 #if defined(MBEDTLS_PLATFORM_C)
 #include "mbedtls/platform.h"
 #else
@@ -409,7 +403,7 @@
 
             memcpy( client_ip, &addr4->sin_addr.s_addr, *ip_len );
         }
-        else
+        /* else
         {
             struct sockaddr_in6 *addr6 = (struct sockaddr_in6 *) &client_addr;
             *ip_len = sizeof( addr6->sin6_addr.s6_addr );
@@ -418,7 +412,7 @@
                 return( MBEDTLS_ERR_NET_BUFFER_TOO_SMALL );
 
             memcpy( client_ip, &addr6->sin6_addr.s6_addr, *ip_len);
-        }
+        } */
     }
 
     return( 0 );
