diff --git a/configure.ac b/configure.ac
index 581596f..44c5167 100644
--- a/configure.ac
+++ b/configure.ac
@@ -122,6 +122,23 @@ AS_IF([test "x$enable_threading" = "xyes" && ! test "$EMSCRIPTEN"], [
   ])
 ])
 
+# FTGL
+AC_ARG_ENABLE([ftgl],
+  AS_HELP_STRING([--enable-ftgl], [FTGL support to display text such as preset title, stats, fps, ...]),
+  [], [enable_ftgl=check])
+AS_IF([test "$enable_ftgl" != "no"], [
+  PKG_CHECK_MODULES([FTGL], [ftgl], [
+    AC_DEFINE([USE_FTGL], [1], [Define USE_FTGL])
+    CFLAGS="$CFLAGS ${FTGL_CFLAGS}"
+    CXXFLAGS="$CXXFLAGS ${FTGL_CFLAGS}"
+    LIBS="$LIBS ${FTGL_LIBS}"
+  ], 
+  [
+    AS_IF([test "$enable_ftgl" = "yes"], AC_MSG_ERROR([*** libftgl not found!]))
+    enable_ftgl=no
+  ])
+])
+
 AC_ARG_ENABLE([gles],
   AS_HELP_STRING([--enable-gles], [OpenGL ES support]),
   [], [enable_gles=no])
@@ -265,6 +282,7 @@ SDL:                    ${enable_sdl}
 Qt:                     ${enable_qt}
 Pulseaudio:             ${enable_pulseaudio}
 Jack:                   ${enable_jack}
+FTGL:                   ${enable_ftgl}
 OpenGLES:               ${enable_gles}
 Emscripten:             ${enable_emscripten}
 llvm:                   ${enable_llvm}