diff -NBaur curl-7.67.0.new/lib/connect.c curl-7.67.0/lib/connect.c
--- curl-7.67.0.new/lib/connect.c	2019-12-03 10:36:21.885160000 -0700
+++ curl-7.67.0/lib/connect.c	2019-12-03 10:37:27.499231800 -0700
@@ -111,6 +111,9 @@
 tcpkeepalive(struct Curl_easy *data,
              curl_socket_t sockfd)
 {
+  (void) sockfd;
+  infof(data, "SO_KEEPALIVE is not supported on 3DS\n");
+#if 0
   int optval = data->set.tcp_keepalive?1:0;
 
   /* only set IDLE and INTVL if setting KEEPALIVE is successful */
@@ -162,6 +165,7 @@
 #endif
 #endif
   }
+#endif
 }
 
 static CURLcode
@@ -494,7 +498,7 @@
 static bool verifyconnect(curl_socket_t sockfd, int *error)
 {
   bool rc = TRUE;
-#ifdef SO_ERROR
+#if 0
   int err = 0;
   curl_socklen_t errSize = sizeof(err);
 
diff -NBaur curl-7.67.0.new/lib/curl_addrinfo.c curl-7.67.0/lib/curl_addrinfo.c
--- curl-7.67.0.new/lib/curl_addrinfo.c	2019-12-03 10:36:22.042114000 -0700
+++ curl-7.67.0/lib/curl_addrinfo.c	2019-12-03 10:37:27.500256100 -0700
@@ -151,9 +151,11 @@
     if((ai->ai_addr == NULL) || !(ai->ai_addrlen > 0))
       continue;
 
+#if 0 // 3DS socu returns ai_addrlen = 8
     /* ignore elements with bogus address size */
     if((size_t)ai->ai_addrlen < ss_size)
       continue;
+#endif
 
     ca = malloc(sizeof(Curl_addrinfo));
     if(!ca) {
diff -NBaur curl-7.67.0.new/lib/socketpair.c curl-7.67.0/lib/socketpair.c
--- curl-7.67.0.new/lib/socketpair.c	2019-12-03 10:36:21.848108500 -0700
+++ curl-7.67.0/lib/socketpair.c	2019-12-03 10:37:27.505455900 -0700
@@ -40,6 +40,7 @@
 #ifdef HAVE_NETINET_IN_H
 #include <netinet/in.h> /* IPPROTO_TCP */
 #endif
+#include <arpa/inet.h>
 #ifndef INADDR_LOOPBACK
 #define INADDR_LOOPBACK 0x7f000001
 #endif /* !INADDR_LOOPBACK */
