diff --git a/src/Makefile b/src/Makefile
index 42bd5c2..f6ae9fb 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -7,16 +7,17 @@ HEADERS = gfx.h gfx_types.h gfx_tex.h gfx_color.h gfx_con.h gfx_con_esc.h
 include ../common.mk
 
 ifeq ($(GAMECUBE),1)
-INSTALLLIB = $(DEVKITPRO)/libogc/lib/cube
+INSTALLLIB = ${DESTDIR}$(DEVKITPRO)/libogc/lib/cube/
 else
-INSTALLLIB = $(DEVKITPRO)/libogc/lib/wii
+INSTALLLIB = ${DESTDIR}$(DEVKITPRO)/libogc/lib/wii/
 endif
 
-INSTALLINCLUDE = $(DEVKITPRO)/libogc/include/gxflux
+INSTALLINCLUDE = ${DESTDIR}$(DEVKITPRO)/libogc/include/gxflux/
 
 install: all
 	@echo "  INSTALL   $(INSTALLLIB)"
-	@install -m 644 $(TARGET) $(INSTALLLIB)
+	@mkdir -p $(INSTALLLIB)
+	@install -m 644 $(TARGET) $(INSTALLLIB)/$(TARGET)
 	@mkdir -p $(INSTALLINCLUDE)
 	@install -m 644 $(HEADERS) $(INSTALLINCLUDE)
 
