diff -r 651485e1d462 common.mk
--- a/common.mk	Sun Sep 29 18:42:52 2019 +0200
+++ b/common.mk	Mon Feb 02 19:05:34 2026 -0300
@@ -31,6 +31,8 @@
 
 DEPDIR = .deps
 
+.PHONY: all
+
 all: $(TARGET)
 
 ifeq ($(LIBRARY),1)
@@ -60,13 +62,15 @@
 	@echo "  STRIP     $@"
 	@$(STRIP) $< -o $@
 
+.PHONY: wiiload geckoload upload
+
 wiiload: $(TARGET_STRIPPED)
 	@echo "  WIILOAD   $<"
-	@$(DEVKITPPC)/bin/wiiload $<
+	@$(DEVKITPRO)/tools/bin/wiiload $<
 
 geckoload: $(TARGET_STRIPPED)
 	@echo "  WIILOAD   $<"
-	@WIILOAD=$(USBGECKODEVICE) $(DEVKITPPC)/bin/wiiload $<
+	@WIILOAD=$(USBGECKODEVICE) $(DEVKITPRO)/tools/bin/wiiload $<
 
 upload: geckoload
 endif
@@ -84,10 +88,10 @@
 	@echo "  ASSEMBLE  $@"
 	@$(CC) $(CFLAGS) $(DEFINES) $(ASFLAGS) -c $< -o $@
 
+.PHONY: clean
+
 clean:
 	@rm -rf $(DEPDIR)
 	@rm -f $(TARGET) $(TARGET_STRIPPED) $(TARGET).map $(OBJS)
 
 -include $(DEPDIR)/*
-
-.PHONY: clean
diff -r 651485e1d462 examples/console/main.c
--- a/examples/console/main.c	Sun Sep 29 18:42:52 2019 +0200
+++ b/examples/console/main.c	Mon Feb 02 19:05:34 2026 -0300
@@ -13,7 +13,13 @@
 
 static bool quit = false;
 
-void stmcb(void) {
+void on_power(void) {
+	quit = true;
+}
+
+void on_reset(u32 irq, void* ctx) {
+	(void)irq;
+	(void)ctx;
 	quit = true;
 }
 
@@ -25,8 +31,8 @@
 	VIDEO_Init();
 	PAD_Init();
 
-	SYS_SetResetCallback(stmcb);
-	SYS_SetPowerCallback(stmcb);
+	SYS_SetResetCallback(on_reset);
+	SYS_SetPowerCallback(on_power);
 
 	gfx_video_init(NULL);
 	gfx_init();
diff -r 651485e1d462 src/Makefile
--- a/src/Makefile	Sun Sep 29 18:42:52 2019 +0200
+++ b/src/Makefile	Mon Feb 02 19:05:34 2026 -0300
@@ -7,18 +7,19 @@
 include ../common.mk
 
 ifeq ($(GAMECUBE),1)
-INSTALLLIB = $(DEVKITPRO)/libogc/lib/cube
+INSTALLLIB = $(DESTDIR)$(DEVKITPRO)/libogc/lib/cube
 else
-INSTALLLIB = $(DEVKITPRO)/libogc/lib/wii
+INSTALLLIB = $(DESTDIR)$(DEVKITPRO)/libogc/lib/wii
 endif
 
-INSTALLINCLUDE = $(DEVKITPRO)/libogc/include/gxflux
+INSTALLINCLUDE = $(DESTDIR)$(DEVKITPRO)/libogc/include/gxflux
+
+.PHONY: install install-all
 
 install: all
 	@echo "  INSTALL   $(INSTALLLIB)"
-	@install -m 644 $(TARGET) $(INSTALLLIB)
-	@mkdir -p $(INSTALLINCLUDE)
-	@install -m 644 $(HEADERS) $(INSTALLINCLUDE)
+	@install -Dm 644 -t $(INSTALLLIB) $(TARGET)
+	@install -Dm 644 -t $(INSTALLINCLUDE) $(HEADERS)
 
 install-all:
 	@$(MAKE) --no-print-directory clean
diff -r 651485e1d462 src/gfx_con.c
--- a/src/gfx_con.c	Sun Sep 29 18:42:52 2019 +0200
+++ b/src/gfx_con.c	Mon Feb 02 19:05:34 2026 -0300
@@ -42,34 +42,38 @@
 static ssize_t _con_read(struct _reent *r, void *fd, char *ptr, size_t len);
 
 static const devoptab_t _dt_stdio = {
-	"stdio",			// device name
-	0,					// size of file structure
-	NULL,				// device open
-	NULL,				// device close
-	_con_write,			// device write
-	_con_read,			// device read
-	NULL,				// device seek
-	NULL,				// device fstat
-	NULL,				// device stat
-	NULL,				// device link
-	NULL,				// device unlink
-	NULL,				// device chdir
-	NULL,				// device rename
-	NULL,				// device mkdir
-	0,					// dirStateSize
-	NULL,				// device diropen_r
-	NULL,				// device dirreset_r
-	NULL,				// device dirnext_r
-	NULL,				// device dirclose_r
-	NULL,				// device statvfs_r
-	NULL,				// device ftrunctate_r
-	NULL,				// device fsync_r
-	NULL,				// deviceData;
-	NULL,				// chmod_r
-	NULL,				// fchmod_r
-	NULL,				// rmdir_r
-	NULL,				// lstat_r
-	NULL,				// utimes_r
+	.name         = "stdio",
+	.structSize   = sizeof(devoptab_t),
+	.open_r       = NULL,
+	.close_r      = NULL,
+	.write_r      = _con_write,
+	.read_r       = _con_read,
+	.seek_r       = NULL,
+	.fstat_r      = NULL,
+	.stat_r       = NULL,
+	.link_r       = NULL,
+	.unlink_r     = NULL,
+	.chdir_r      = NULL,
+	.rename_r     = NULL,
+	.mkdir_r      = NULL,
+	.dirStateSize = 0,
+	.diropen_r    = NULL,
+	.dirreset_r   = NULL,
+	.dirnext_r    = NULL,
+	.dirclose_r   = NULL,
+	.statvfs_r    = NULL,
+	.ftruncate_r  = NULL,
+	.fsync_r      = NULL,
+	.deviceData   = NULL,
+	.chmod_r      = NULL,
+	.fchmod_r     = NULL,
+	.rmdir_r      = NULL,
+	.lstat_r      = NULL,
+	.utimes_r     = NULL,
+	.fpathconf_r  = NULL,
+	.pathconf_r   = NULL,
+	.symlink_r    = NULL,
+	.readlink_r   = NULL,
 };
 
 static const GXColor _con_colors[] = {
@@ -312,6 +316,7 @@
 			default:
 				break;
 			}
+			break;
 
 		case 'm': // Set Attribute Mode
 			if (!count)
