--- a/ode/src/threading_pool_posix.cpp
+++ b/ode/src/threading_pool_posix.cpp
@@ -371,7 +371,6 @@
 
         int set_result;
         if ((set_result = pthread_attr_setdetachstate(thread_attr, PTHREAD_CREATE_JOINABLE)) != EOK
-            || (set_result = pthread_attr_setinheritsched(thread_attr, PTHREAD_INHERIT_SCHED)) != EOK
 #if (HAVE_PTHREAD_ATTR_SETSTACKLAZY)
             || (set_result = pthread_attr_setstacklazy(thread_attr, PTHREAD_STACK_NOTLAZY)) != EOK
 #endif
