diff -NBaur libxmp-4.4.1-old/src/load.c libxmp-4.4.1-new/src/load.c
--- libxmp-4.4.1-old/src/load.c	2016-10-12 11:30:03.000000000 +0100
+++ libxmp-4.4.1-new/src/load.c	2019-09-22 21:13:57.155910284 +0100
@@ -82,6 +82,7 @@
 
 static int execute_command(char *cmd, char *filename, FILE *t)
 {
+#if defined(HAVE_POPEN) || defined(WIN32)
 	char line[1024], buf[BUFLEN];
 	FILE *p;
 	int n;
@@ -111,7 +112,7 @@
 	}
 
 	pclose (p);
-
+#endif
 	return 0;
 }
 
@@ -171,7 +172,7 @@
 		return 0;
 	}
 
-#if defined __ANDROID__ || defined __native_client__
+#if defined __ANDROID__ || defined __native_client__ || !defined(HAVE_POPEN)
 	/* Don't use external helpers in android */
 	if (cmd) {
 		return 0;
