--- CMakeLists.txt.orig	2022-04-11 22:55:49.000000000 +0000
+++ CMakeLists.txt	2023-09-11 23:40:47.113419200 +0000
@@ -30,7 +30,7 @@
 option(AVIF_ENABLE_WERROR "Treat all compiler warnings as errors" ON)
 
 option(AVIF_CODEC_AOM "Use the AOM codec for encoding/decoding (see AVIF_CODEC_AOM_DECODE/AVIF_CODEC_AOM_ENCODE)" OFF)
-option(AVIF_CODEC_DAV1D "Use the dav1d codec for decoding" OFF)
+option(AVIF_CODEC_DAV1D "Use the dav1d codec for decoding" ON)
 option(AVIF_CODEC_LIBGAV1 "Use the libgav1 codec for decoding" OFF)
 option(AVIF_CODEC_RAV1E "Use the rav1e codec for encoding" OFF)
 option(AVIF_CODEC_SVT "Use the SVT-AV1 codec for encoding" OFF)
