From 59e0d7596fc199959b5dd631b4857d351cc0aa5f Mon Sep 17 00:00:00 2001
From: TuxSH <1922548+TuxSH@users.noreply.github.com>
Date: Tue, 24 May 2022 19:51:25 +0100
Subject: [PATCH] Fix PMAPP_TerminateTitle

---
 libctru/source/services/pmapp.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libctru/source/services/pmapp.c b/libctru/source/services/pmapp.c
index 2e3bca2..ec704e5 100644
--- a/libctru/source/services/pmapp.c
+++ b/libctru/source/services/pmapp.c
@@ -79,7 +79,7 @@ Result PMAPP_TerminateTitle(u64 titleId, s64 timeout)
 	Result ret = 0;
 	u32 *cmdbuf = getThreadCommandBuffer();
 
-	cmdbuf[0] = IPC_MakeHeader(0x3, 4, 0); // 0x40100
+	cmdbuf[0] = IPC_MakeHeader(0x4, 4, 0); // 0x40100
 	cmdbuf[1] = (u32)titleId;
 	cmdbuf[2] = (u32)(titleId >> 32);
 	cmdbuf[3] = (u32)timeout;
-- 
2.36.1.windows.1

