From 4866df79a5d929c211433942f13c80b5996bb9d9 Mon Sep 17 00:00:00 2001
From: fincs <fincs@devkitpro.org>
Date: Thu, 23 Jul 2020 02:45:39 +0200
Subject: [PATCH] condvar: fix embarrassing typo that prevented it from
 actually working

---
 libctru/source/synchronization.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/libctru/source/synchronization.c b/libctru/source/synchronization.c
index 48c9fecf..fe2ade3f 100644
--- a/libctru/source/synchronization.c
+++ b/libctru/source/synchronization.c
@@ -159,7 +159,7 @@ static inline void CondVar_BeginWait(CondVar* cv, LightLock* lock)
 	s32 val;
 	do
 		val = __ldrex(cv) - 1;
-	while (!__strex(cv, val));
+	while (__strex(cv, val));
 	LightLock_Unlock(lock);
 }
 
@@ -180,7 +180,7 @@ static inline bool CondVar_EndWait(CondVar* cv, s32 num_threads)
 			else
 				val = 0;
 		}
-	} while (!__strex(cv, val));
+	} while (__strex(cv, val));
 
 	return hasWaiters;
 }
-- 
2.11.0

