diff -Naur libtheora-1.2.0alpha1.orig/lib/arm/armcpu.c libtheora-1.2.0alpha1/lib/arm/armcpu.c
--- libtheora-1.2.0alpha1.orig/lib/arm/armcpu.c	2010-09-23 13:21:44.000000000 -0700
+++ libtheora-1.2.0alpha1/lib/arm/armcpu.c	2021-03-18 15:13:29.976897091 -0700
@@ -107,6 +107,22 @@
   return flags;
 }
 
+#elif defined(_3DS)
+
+ogg_uint32_t oc_cpu_flags_get(void){
+  ogg_uint32_t  flags;
+  flags=0;
+
+#if defined(OC_ARM_ASM_EDSP)
+  flags|=OC_CPU_ARM_EDSP;
+#endif
+#if defined(OC_ARM_ASM_MEDIA)
+  flags|=OC_ARM_ASM_MEDIA;
+#endif
+
+  return flags;
+}
+
 #else
 /*The feature registers which can tell us what the processor supports are
    accessible in priveleged modes only, so we can't have a general user-space

